trace/client/src/main/scala/org/http4s/otel4s/middleware/trace/client/AttributeProvider.scala

```diff
@@ -0,0 +1,81 @@
+/*
+ * Copyright 2023 http4s.org
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.http4s
+package otel4s.middleware
+package trace.client
+
+import org.typelevel.ci.CIString
+import org.typelevel.otel4s.Attributes
+
+/** Provides attributes for spans using requests and responses.
+  *
+  * @note Implementations MUST NOT access request or response bodies.
+  */
+trait AttributeProvider { self =>
+
+  /** Provides attributes for a span using the given request.
+    *
+    * @note Implementation MUST NOT access request body.
+    */
+  def requestAttributes[F[_]](
+      request: Request[F],
```

Comment:
Would `RequestPrelude` be enough to fill out all necessary attributes?

Reply:
no, see [here](https://github.com/http4s/http4s-otel4s-middleware/pull/158/files#diff-653af73ff5e9e8953dbcdd9e6dc9a019698f624cfb939e6dfffd61b512866252R187) and [here](https://github.com/http4s/http4s-otel4s-middleware/pull/158/files#diff-653af73ff5e9e8953dbcdd9e6dc9a019698f624cfb939e6dfffd61b512866252R196) for example

Reply:
I will note that the actual implementation that calls these methods replaces the message body with `Stream.empty` first for safety regardless

Reply:
Would URI authority `request.uri.authority` work? https://github.com/http4s/http4s/pull/7491#issuecomment-2400187857
I haven't tried it myself yet, so I'm unsure, to be honest.

Reply:
I honestly don't know why `RequestPrelude` doesn't include `attributes`, it's not like they're mutable

Reply:
Attributes in RequestPrelude would really help there, too: https://github.com/http4s/http4s/pull/7491.
Any chance we can make this happen?

Reply:
... no :( from the scaladocs

> The [[Request#attributes]] are omitted in this encoding because they
> do not (and can not) have a [[cats.kernel.Order]] instance. If they
> were included here, then we could not write a [[cats.kernel.Order]]
> instance for [[RequestPrelude]], limiting some of its utility, e.g. it
> could not be used in a [[cats.data.NonEmptySet]].

same for `ResponsePrelude`. I think this is silly, but folks will probably object to breaking that rather useless property